torch/_inductor/codecache.py

```diff
@@ -575,79 +588,212 @@ def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
     """
     Supports caching and reusing compiled Fx graphs.
+
+    The overall strategy is as follows:
+    - This cache stores entries on disk. When saving an entry, we can't
+      serialize callables (that could be C++, Triton, etc.), so we serialize
+      their own disk cache location. We then recreate the compiled artifact
+      after fetching from disk.
+    - For indexing the cache, we gather the fields relevant to identifying an
+      FxGraph (the graph module, graph inputs, system settings etc.) into an
+      FxGraphCacheDetails object, pickle it, and compute a hash for the key.
+      See FxGraphCachePickler.
+    - Among the metadata we store, we also include a guards expression that's
+      appropriate for validating any symbols for Tensor arguments that have
+      symbolic bounds. On cache lookup then, we evaluate those guards in the
+      current context to validate that a cached entry can be served.
+    - A given graph could have multiple compiled versions, corresponding to
+      different sets of guards. Therefore, we store cache entries in the form:
+          <temp dir>/<fx graph hash>/<serialized metatdata>
+    - On lookup, we compute the key from the graph details, iterate over all
+      leaf files in the corresponding subdirectory, deserialize the entry, and
+      evaluate its guards expression. If the evaluation succeeds, we have a
+      cache hit. If it fails, we compile the graph and store a new entry.
+    - Finally, on a cache hit, we need to make sure any guards that would
+      have been created during compilation are added to the current context.
     """
 
     # TODO(masnesral): Investigate whether it's beneficial to store compiled graphs
     # in an in-memory cache after loading from disk.
-    @classmethod
-    def save_graph(cls, key: str, compiled_graph: CompiledFxGraph):
+    @staticmethod
+    def _get_subdir(key: str) -> str:
+        """
+        Return the disk location for a given cache key.
+        """
+        return os.path.join(cache_dir(), key[1:3], key)
+
+    @staticmethod
+    def _filter_symints(inputs: List[Any]) -> List[torch.SymInt]:
+        """
+        Get the SymInt objects from the input list.
+        """
+        return [s for s in inputs if isinstance(s, torch.SymInt)]
+
+    @staticmethod
+    def _get_shape_env() -> ShapeEnv:
+        """
+        Helper to get the shape env from the tracing context.
+        """
+        tracing_context = torch._guards.TracingContext.get()
+        assert tracing_context is not None
+        return tracing_context.fake_mode.shape_env
+
+    @staticmethod
+    def _lookup_graph(
+        key: str,
+        example_inputs: List[torch.Tensor],
+    ) -> Optional[CompiledFxGraph]:
+        """
+        Lookup a compiled graph in the cache by key. On a hit, return the
+        deserialized CompiledFxGraph object. On a miss, return None.
+        """
+        subdir = FxGraphCache._get_subdir(key)
+        if not os.path.exists(subdir):
```

Comment:
This function needs more logging.

Reply:
In particular, eventually we're going to be debugging weird production issues related to cache related problems, and I need 100% of the relevant file system state we probe to be reported by the logs here, because it's going to be extremely inconvenient to actually get to the machine that has the problem

Reply:
@ezyang it's related to the discussion above about using an env var. I assumed that dumping all this info is going to waaay too verbose for normal debug output -- because it's a lot. Lemme know if that assumption is wrong and I'll just log.debug it. I haven't investigated what you mean by using an artifact logger, but does artifact logging the info give you the control you need?

Reply:
Yes. You can configure artifact logging so it won't get triggered if you do `+inductor`. But I think that there is probably a non-noisy version of the logging which is acceptable on `+inductor` by default. If you are O(1) log lines per inductor compilation that is DEFINITELY acceptable